src/bzz-link.ts

```diff
@@ -23,20 +27,100 @@ export function requestFilter(pathname: string, req: Request): boolean {
   return req.subdomains.length >= 1
 }
 
+export function getDomain(req: Request): string | undefined {
+  try {
+    const matches = req.headers.host!.match(/^([^?#]*)(\?([^#]*))?(#(.*))?:[0-9]*]?/i) // clean host domain string
+    const match = matches && matches[1] // domain will be null if no match is found
+
+    if (match) return match
+  } catch {
+    return req.headers.host
+  }
+}
+
+export function validateDomain(host: string, domains: string[]): boolean {
+  return domains.includes(host)
+}
+
+/**
+ * Function that search for DNSLink configuration on TXT records
+ *
+ * @param domain
+ */
+async function dnsLookup(
+  domain: string,
+  queryDnsLinkEndpoint = DEFAULT_QUERY_DNSLINK_ENDPOINT,
+): Promise<Result | undefined> {
+  try {
+    return await resolve(domain, {
+      endpoints: [queryDnsLinkEndpoint],
+      timeout: 1000, // timeout for the operation
+      retries: 3, // retries in case of transport error
+    })
+  } catch (ex) {
+    logger.error(`dnslink lookup error`, ex)
+    throw new NoDNSLinkFoundError(`dnslink lookup error resolving domain ${domain}`)
+  }
+}
+
+function getDnslinkBzzRoute(hostname: string, target: string, result: Result | undefined): string | undefined {
+  if (result) {
+    const { txtEntries } = result
+    const txtEntry = JSON.parse(JSON.stringify(txtEntries[0]))
+
+    if (txtEntry.value) {
+      const bzzResource = txtEntry.value.split('/')[2]
+      logger.info(`bzz link proxy`, { hostname, bzzResource })
+
+      return getBzzRoute(target, bzzResource)
+    }
+  }
+}
+
+function getBzzRoute(target: string, bzzResource: string): string {
+  return `${target}/bzz/${bzzResource}`
+}
+
 /**
- * Closure that routes subdomain CID/ENS to /bzz endpoint.
+ * Closure that routes subdomain CID/ENS/DNSLINK to /bzz endpoint.
  *
  * @param target
+ * @param dnslinkDomains
  * @param isCidEnabled
  * @param isEnsEnabled
+ * @param isDnslinkEnabled
+ * @param dnsQuery
  */
-export function routerClosure(target: string, isCidEnabled: boolean, isEnsEnabled: boolean) {
-  return (req: Request): string => {
-    const bzzResource = subdomainToBzz(req, isCidEnabled, isEnsEnabled)
+export function routerClosure(
+  target: string,
+  dnslinkDomains: string[],
+  isCidEnabled: boolean,
+  isEnsEnabled: boolean,
+  isDnslinkEnabled: boolean,
+  dnsQuery?: string,
+): { (req: Request): Promise<string> } {
+  return async (req: Request): Promise<string> => {
+    if (isDnslinkEnabled) {
+      let domain = getDomain(req)
+
+      // in case domain is not on the list of allowed domains then it uses anything on headers-host
+      if (!(domain && validateDomain(domain, dnslinkDomains))) {
```

Comment:
This does not make sense to me.  If the domain is not listed in valid domains then no DNSLink should be looked up, no?